File: src/com/OpenSerenity/utils/Waiter.java
package com.OpenSerenity.utils;

import com.OpenSerenity.core.TestContext;
import com.OpenSerenity.exceptions.TimeoutException;
import com.OpenSerenity.functionalInterfaces.WaitCondition;

import java.util.Date;

public class Waiter {
    public static Waiter Default() {
        return new Waiter();
    }

    public static Waiter WithTimeout(long timeout) {
        return new Waiter(timeout);
    }

    private Waiter() {
    }

    private Waiter(long timeout) {
        this.timeout = timeout;
    }

    long timeout = TestContext.configuration.getTimeout();
    long retryTimeout = TestContext.configuration.getRetryTimeout();

    public void waitFor(WaitCondition waitCondition) throws Exception {
        try {
            long start = new Date().getTime();
            do {
                try {
                    if (waitCondition.invoke()) {
                        return;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                } finally {
                    sleepFor(retryTimeout);
                }
            } while (!timeoutReached(start));
        } catch (Exception e) {
            e.printStackTrace();
        }
        throw new TimeoutException();
    }

    private boolean timeoutReached(long start) {
        return (new Date().getTime() - start + retryTimeout) > timeout;
    }

    private void sleepFor(long millis) throws InterruptedException {
        Thread.sleep(millis);
    }
}

File: src/scala/com/OpenSerenity/utils/Waiter.scala
package com.OpenSerenity.utils

import com.OpenSerenity.core.TestContext
import java.util.Date
import com.OpenSerenity.exceptions.TimeoutException


class Waiter(timeout: Long, retryTime: Long) {
  def this() = this(TestContext.configuration.getTimeout, TestContext.configuration.getRetryTimeout)
  def this(timeout: Long) = this(timeout, TestContext.configuration.getRetryTimeout)


  def waitFor(waitCondition: () => Boolean): Unit = {
    try {
      val start: Long = new Date().getTime()
      do {
        try {
          if (waitCondition()) return
        } catch {
          case e: Exception => e.printStackTrace()
        } finally {
          sleepFor(retryTime)
        }
      } while (!timeoutReached(start))
    } catch {
      case e: Exception => e.printStackTrace()
    }

    throw new TimeoutException()
  }

  private def timeoutReached(start: Long): Boolean = (new Date().getTime() - start + retryTime) > timeout
  private def sleepFor(millis: Long) = Thread.sleep(millis)
}

object Waiter {
  def default = new Waiter()
  def withTimeout(_timeout: Long) = new Waiter(timeout = _timeout)
}
